### PR TITLE
Enable FindBugs in the plugin, generate XML reports

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] Update FindBugs version
- [x] Enable XML report generation
- [x] Enable reporting on https://jenkins.ci.cloudbees.com/job/plugins/job/promotes-builds-plugin/
- [x] Cleanup the code if possible

The code is not clean according to the local run

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@reviewbybees
The code should be clean now

File: pom.xml [hunk 1]
```diff
@@ -13,6 +13,12 @@
 
   <name>Jenkins promoted builds plugin</name>
   <url>http://wiki.jenkins-ci.org/display/JENKINS/Promoted+Builds+Plugin</url>
+  
+  <properties> 
+    <findbugs-maven-plugin.version>3.0.1</findbugs-maven-plugin.version>
+    <!--TODO: change to true after the code cleanup-->
+    <findbugs-maven-plugin.failOnError>true</findbugs-maven-plugin.failOnError>
```
Comment (stephenc) [inline pom.xml]: :bug: property name

File: pom.xml [hunk 2]
```diff
@@ -71,6 +83,25 @@
         <artifactId>maven-release-plugin</artifactId>
         <version>2.5.1</version>
       </plugin>
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>findbugs-maven-plugin</artifactId>
+        <version>${findbugs-maven-plugin.version}</version>
+        <configuration>    
+          <failOnError>${findbugs-maven-plugin.failOnError}</failOnError>
```
Comment (stephenc) [inline pom.xml]: :bug: property name

File: src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java [hunk 3]
```diff
@@ -251,7 +252,9 @@ public PromotionProcess createProcessFromXml(final String name, InputStream xml)
         }
         File configXml = Items.getConfigFile(getRootDirFor(name)).getFile();
         File dir = configXml.getParentFile();
-        dir.mkdirs();
+        if (!dir.mkdirs()) {
+            throw new IIOException("Cannot create directories for "+dir);
```
Comment (stephenc) [inline src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java]: `IIOException`??? :bug:

Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/JobPropertyImpl.java]: Damn autocompletion + typos :(

**Comment by stephenc** [comment]
:bug: wrong exception and bad property name

**Comment by oleg-nenashev** [comment]
@stephenc Fixed both comments

**Comment by stephenc** [comment]
:bee: though I do not see the reason behind introducing the property when it is used once and only once

**Comment by oleg-nenashev** [comment]
@reviewbybees done
Squashing the commits

**Comment by oleg-nenashev** [comment]
@dnozay 
Seems you're the plugin's maintainer in-fact. Do you agree with merging this change?

File: src/main/java/hudson/plugins/promoted_builds/PromotionCause.java [hunk 4]
```diff
@@ -53,4 +53,40 @@ public void print(final TaskListener listener) {
         );
     }
     // }}}
+
+    @Override
+    public int hashCode() {
+        int hash = 7;
```
Comment (dnozay) [inline src/main/java/hudson/plugins/promoted_builds/PromotionCause.java]: minor: i'm not familiar with how you make a hash function good enough. adding comments would be great.

**Comment by dnozay** [comment]
@oleg-nenashev - i'm going to be AFK for about 10 days. change looks fine to merge in.

**Comment by oleg-nenashev** [comment]
Merging the change to enforce FindBugs in other pending PRs

**Pull request merged**